### ISSUE TITLE
Part 4. Gradient Descent

### ISSUE THREAD

**Issue by XuSShuai** [XuSShuai/Machine-Learning-2017-Fall#4]
## Review: Gradient Descent

在机器学习的第三步，我们需要求解一个包含模型参数的最优化问题：
$$\theta^{*} = argmax_{\theta} \mathcal{L}(\theta)$$
假设$\theta = \\{\theta_1, \theta_2\\}$
$$\bigtriangledown L(\theta) = 
\left[
\begin{matrix}
\frac{\partial L}{\partial \theta_1}  \\\\
\frac{\partial L}{\partial \theta_2}
\end{matrix}
\right] \tag{3}
$$
梯度下降过程：
$$
\theta^1 = \theta^0 - \eta \bigtriangledown L(\theta^{0}) \\\\
\theta^2 = \theta^1 - \eta \bigtriangledown L(\theta^{1}) \\\\
\cdots
$$
![image](https://user-images.githubusercontent.com/29543183/48822600-97a84780-ed98-11e8-86e4-3500007ed09c.png)

## Learning Rate
**set the learning rate carefully.**
![image](https://user-images.githubusercontent.com/29543183/48822748-143b2600-ed99-11e8-8fbe-12822e47eb7a.png)

## Adaptive Learning Rate
Popular & Simple idea: reduce the learning rate by some factor every few epochs, e.g. $\eta^{t} = \frac{\eta}{\sqrt{t + 1}}$
**But
learning rate cannot be one-size-fit-all, giving different parameters different learning is better !**

## Adagrad

每一个参数的learning rate除以该参数之前所有偏微分的平方和的开方。所以每一个参数的learning rate都是不同的。

### Adagrad的计算过程

$$
w^1 = w^0 - \frac{\eta^0}{\sigma^0}g^{0} ,\ \sigma^0 = \sqrt{(g^0)^2} \\\\
w^2 = w^1 - \frac{\eta^1}{\sigma^1}g^{1} ,\ \sigma^1 = \sqrt{\frac{1}{2}((g^0)^2+(g^1)^2)} \\\\
w^3 = w^2 - \frac{\eta^2}{\sigma^2}g^{2} ,\ \sigma^2 = \sqrt{\frac{1}{3}((g^0)^2+(g^1)^2+(g^1)^2)} \\\\
\cdots \\\\
w^{t+1} = w^{t}  - \frac{\eta^t}{\sigma^t}g^{t} ,\ \sigma^t = \sqrt{\frac{1}{t+1}\sum_{i}^{t}(g^i)^2}
$$
where
$$\eta^{t} = \frac{\eta}{\sqrt{t+1}}, g^{t} = \frac{\partial L(w^t)}{\partial w}$$

所以

$$w^{t+1} = w^{t}  - \frac{\eta^t}{\sigma^t}g^{t} =  w^{t}  - \frac{\frac{\eta}{\sqrt{t+1}}}{\sqrt{\frac{1}{t+1}\sum_{i}^{t}(g^i)^2}}g^{t} = w^{t} - \frac{\eta}{\sum_{i}^{t}(g^i)^2}g^{t}$$
